Workshop: Components 2 (#156)
CodeYourFuture/CYF-Workshops: components-2/readme.md

- [ ] 👪🏿 Collaborate to build a page
```

<!--{{<note type="activity" title="Set up your working branch 15m">}}-->

### Set up your working branch

Everyone should do this so everyone is set up correctly.

1. You will be working on a special branch called [FOR-USE-IN-CLASS](https://github.com/CodeYourFuture/CYF-Workshops/tree/FOR-USE-IN-CLASS).
1. First, **clone** [the workshops repository](https://github.com/CodeYourFuture/CYF-Workshops/)
1. Checkout `git checkout FOR-USE-IN-CLASS`
1. Your cohort has a named _sub branch_. Everyone, check that out next with `git checkout COHORT` (eg `NW6`). This is your working branch:

```mermaid
gitGraph
    commit "main"
    branch FOR-USE-IN-CLASS
    checkout FOR-USE-IN-CLASS
    branch NW6
    checkout NW6
    branch feature/header
    commit
    commit
    checkout NW6
    branch feature/footer
    commit
    commit
    commit
    checkout NW6
    branch feature/card
    commit
    commit
    commit
    checkout NW6
    branch feature/menu
    commit
    commit
    commit
    checkout NW6
    merge feature/header
    merge feature/footer
    merge feature/card
    merge feature/menu
```

> Check you are in the right place with GitLens!

Reviewer: I think I would move this down to the point where they choose a driver and confirm they're on the right branch.

<!--{{</note>}}-->

<!--{{<note type="tip" title="Your working branch is your cohort name">}}-->

Your COHORT branch is your cohort name. It might be NW6, WM5, GL7...You're working in your COHORT branch and NOT in main.

<!--{{</note>}}-->

> Next, divide the class into four teams. Use the [team generator](teams.html) to help you do this.

Reviewer: ```suggestion
Next, divide the class into four teams: header, footer, card, menu. Use the [team generator](teams.html) to help you do this.
```

Is it worth noting that this link seems to be broken in the components-1 workshop when on the curriculum site?

Set a whole class [timer for 10 minutes](https://www.google.com/search?q=10+minute+time).

### Each team, choose a driver

> 1. Driver, ask everyone in your team to confirm you are in the right working branch.

Reviewer: ```suggestion
1. Driver, ask everyone in your team to confirm you are in the right working branch. Tip: try using GitLens!
```

> 1. Driver, from this branch, create and checkout a new branch for your team named like this: `feature/component`, where _cohort_ is your cohort name and _component_ is the component you're building, for example: `feature/header`

Reviewer: ```suggestion
1. Driver, from this branch, create and switch to a new branch for your team named like this: `feature/cohort-component`, where _cohort_ is your cohort name and _component_ is the component you're building, for example: `feature/nw6-header`
```

```bash
git switch --create feature/cohort-component
```

Now you're all set up to make changes to together.html.

<!--{{</note>}}-->

<!--{{<note type="activity" title="Find your docs 10m">}}-->

## Find your docs

Set a whole class [timer for 10 minutes](https://www.google.com/search?q=10+minute+time).

Deployed from this repo is a [mini design system](https://cyf-workshops.netlify.app/components-2). This design system has several documented components, and some also have javascript files.

> Your mission is to find the docs for each component with an associated script file, and read them. Discuss your findings. Look at the pages on the website with Devtools. Compare the Inspect view with the source code you can see in the files. How are they different? How are HTML and the DOM different?

Reviewer: So I think the intention is that trainees figure out that they need to import the script and call the "component" - is that right?

Author: Right, all the pages are built in this way so they can just copy and paste it from index.html or any doc page.

<!--{{</note>}}-->

> 
> ### Together separately
> 
> Set a whole class [timer for 30 minutes](https://www.google.com/search?q=30+minute+time).
> 
> <!--{{<note type="activity" title="Parallel Development 30m">}}-->
> 

Reviewer: For consistency:

```suggestion

<!--{{<note type="activity" title="Parallel Development 30m">}}-->

### Together separately

Set a whole class [timer for 30 minutes](https://www.google.com/search?q=30+minute+time).
```

Now **build** your component. Elect a new driver.

1. Find [together.html](together.html) in the /components-2 directory. Your mission is to complete this page.
1. In the correct portion of the page, make your changes to build _your feature_ on together.html.
1. Driver, commit your changes to your feature branch.
1. Open a pull request from your branch to the COHORT branch, with your changes. Ask for help!
1. Remember, docs will help you too.
1. Do the simplest thing that could possibly work to begin with. You can always update your PR with more changes.

```mermaid
gitGraph
    commit id: "start"
    branch feature/component
    commit id: "import component function"
    commit id: "import data"
    commit id: "lighthouse revisions"
    checkout COHORT
    merge feature/component

> ```

Reviewer: I'm not a mermaid expert but this seems to be throwing an error?

Author: oh it should be 

```mermaid
gitGraph
    commit id: "start"
    branch FOR-USE-IN-CLASS
    branch COHORT
    branch feature/component
    commit id: "import component function"
    commit id: "import data"
    commit id: "lighthouse revisions"
    checkout COHORT
    merge feature/component
  ```
  
  Somehow I didn't notice this despite having the mermaid vscode installed! TY

<!--{{</note>}}-->

### Review and merge

Now we're going to review each other's work and merge to our cohort branch.

<!--{{<note type="warning" title="Your working branch is your cohort name">}}-->

> Your COHORT branch is your cohort name. You're working in your COHORT branch and NOT in main. Open your pull request to your COHORT branch. Don't open PRs to main.

Reviewer: ```suggestion
Your COHORT branch is your cohort name. You're working in your COHORT branch and NOT in main. Open your pull request to your COHORT branch.

Don't open PRs to main. If you do open a PR to main and it is merged then it would break the exercise for other trainees.
```

Author: I also beefed up the PR approvals to 2 on this repo to reduce this risk.

<!--{{</note>}}-->

Set a whole class [timer for 30 minutes](https://www.google.com/search?q=30+minute+time).

<!--{{<note type="activity" title="Review 30m">}}-->

1. Go to the [pull requests](https://github.com/CodeYourFuture/CYF-Workshops/pulls) for this repo.
1. Either individually or as a group, review each other's PRs.
1. Once you're all happy with the PR, merge it.

> 1. Did it work? Why/why not? Discuss.

Reviewer: Do we need to more clearly define what "worked" actually means?

I guess my underlying question is whether trainees will consider "clicking the merge button" as "working". So we might need to put in something about checking the outcome in `together.html`

But I can also see that coming out of the discussion, especially if guided by volunteers.

Author: I think we can bear a line prodding them to look at together.html  😉
